manager/src/main/java/org/openremote/manager/alarm/AlarmResourceImpl.java

```diff
@@ -0,0 +1,183 @@
+        }
+
+        return alarm.getRealm();
+    }
+
+    private Set<String> validateExistingAlarmIds(List<Long> alarmIds) {
+        alarmIds.forEach(this::validateAlarmId);
+
+        List<SentAlarm> alarms = alarmService.getAlarms(alarmIds);
+
+        if (alarms == null || alarms.size() != alarmIds.size()) {
+            throw new WebApplicationException("One or more alarms do not exist", BAD_REQUEST);
+        }
+
+        Set<String> realms = alarms.stream().map(SentAlarm::getRealm).collect(Collectors.toSet());
+
+        if (realms.stream().anyMatch(realm -> !isRealmActiveAndAccessible(realm))) {
+            throw new WebApplicationException("One or more alarms are in a nonexistent, inactive or inaccessible realm", FORBIDDEN);
+        }
+
+        return realms;
+    }
+
+    private void validateRealm(String realm) {
+        if (!isRealmActiveAndAccessible(realm)) {
+            throw new WebApplicationException("Realm '" + realm + "' is not active or inaccessible", Status.FORBIDDEN);
+        }
+    }
+
+    private void validateRealms(Stream<String> realms) {
+        realms.forEach(this::validateRealm);
+    }
+
+    private List<SentAlarm> filterByActiveAndAccessibleRealms(List<SentAlarm> sentAlarms) {
+        return sentAlarms.stream().filter(sa -> isRealmActiveAndAccessible(sa.getRealm())).toList();
+    }
+
+    @Override
+    public SentAlarm[] getAlarms(RequestParams requestParams, String realm, Alarm.Status status, String assetId, String assigneeId) {
+        String filterRealm = TextUtil.isNullOrEmpty(realm) ? getAuthenticatedRealm().getName(): realm;
+        validateRealm(filterRealm);
+        return alarmService.getAlarms(filterRealm, status, assetId, assigneeId).toArray(new SentAlarm[0]);
+    }
+
+    @Override
+    public void removeAlarms(RequestParams requestParams, List<Long> alarmIds) {
+        Set<String> realms = validateExistingAlarmIds(alarmIds);
+        alarmService.removeAlarms(alarmIds, realms);
+    }
+
+    @Override
+    public SentAlarm getAlarm(RequestParams requestParams, Long alarmId) {
+        validateExistingAlarmId(alarmId);
+        return alarmService.getAlarm(alarmId);
+    }
+
+    @Override
+    public void removeAlarm(RequestParams requestParams, Long alarmId) {
+        String realm = validateExistingAlarmId(alarmId);
+        alarmService.removeAlarm(alarmId, realm);
+    }
+
+    @Override
+    public SentAlarm createAlarm(RequestParams requestParams, Alarm alarm) {
+        validateRealm(alarm.getRealm());
+        try {
+            if (getUserId() != null) {
+                alarm.setSource(MANUAL);
+                alarm.setSourceId(getUserId());
+            }
+            else if (getClientId() != null) {
+                alarm.setSource(CLIENT);
+                alarm.setSourceId(getClientId());
+            }
+            return alarmService.sendAlarm(alarm);
```

Comment:
I don't think it will ever return `null`. `persistenceService.doReturningTransaction` can throw an exception which would result in a "500 Internal Server Error".